server/src/main/java/org/opensearch/transport/OutboundHandler.java

```diff
@@ -83,14 +74,16 @@ final class OutboundHandler {
     ) {
         this.nodeName = nodeName;
         this.version = version;
-        this.features = features;
         this.statsTracker = statsTracker;
         this.threadPool = threadPool;
-        this.bigArrays = bigArrays;
+        this.protocolMessageHandlers = Map.of(
```

Comment:
@VachaShah I think we need to think a bit more how separate that. Here is the idea which I think should work. 

The `OutboundHandler` should be in sync with `InboundHandler` with respect to the protocol: if the inbound message is in protobuf, than the outbound should also be in protocol. There are 2 issues with `OutboundHandler` (at least, as I see it):
 - it is created before `InboundHandler` (so we cannot make those agree on protocol)
 - `OutboundHandler` is mixing 2 things now: sending raw bytes and transport requests / responses (which involves serialization that we would need to change)

The first thing we need to do I believe is to strip from `OutboundHandler` everything which is related to transport requests / responses: basically that would leave as with `void sendBytes(TcpChannel channel, BytesReference bytes, ActionListener<Void> listener)` only. 

Now, the `OutboundHandler` is passed to `InboundHandler` and consequently to `NativeMessageHandler` for now (and protobuf later). At this point things become more straightforward:
 - `NativeMessageHandler` would create an instance of `NativeOutboundHandler`  that would implement the  `sendResponse` / `sendErrorResponse` / `sendRequest` using native protocol and send out raw bytes using the  `OutboundHandler`
 -  consequently, `ProtoobufMessageHandler` would create an instance of `ProtoobufOutboundHandler`  that would implement the  `sendResponse` / `sendErrorResponse` / `sendRequest` using protobuf protocol and send out raw bytes using the  `OutboundHandler`

I think that should work, what do you think?

Reply:
Interesting! I always thought that `InboundHandler` should be in sync with `OutboundHandler`, so if the outbound message is in protobuf, inbound should also be protobuf.

When we strip everything from `OutboundHandler` except `sendBytes`, are we essentially making `TransportResponse`, `TransportRequest` specific to native protocol? And when protobuf is introduced, it would have its own version of `TransportResponse`?

Reply:
> Interesting! I always thought that InboundHandler should be in sync with OutboundHandler

Correct, but inbound comes first , right? (this is the request side)

> are we essentially making TransportResponse, TransportRequest specific to native protocol? And when protobuf is introduced, it would have its own version of TransportResponse?

The serdes of TransportRequest  / TransportResponse is protocol specific, yes.

Reply:
Makes sense! Thank you @reta. Apologies for the delay in response, I was out over the weekend. Let me look into how we can achieve the above idea and I will add an update on this PR once that is done.

Reply:
Just an update here: `sendBytes` method in `OutboundHandler` is not used by general messages, it is made specially for keep alive messages I think because for `TransportResponse` and such messages, they are serialized to bytes using `MessageSerializer` and sent over the wire so thats a separate flow.

Reply:
I was looking into this implementation and the problem I faced is that the `OutboundHandler` and `InboundHandler` combination is decided at [`TcpTransport`](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/transport/TcpTransport.java#L191-L208) class itself. Now I could make `OutboundHandler` as an interface which `NativeOutboundMessageHandler` would implement and the `NativeMessageHandler` can create an instance of this specific native outbound handler ([here](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/transport/NativeMessageHandler.java#L95)). But the problem lies in `TcpTransport` where the outbound handler is also used. So we would need a specific implementation here as well and not an interface. Also, as I mentioned in the previous comment, `sendBytes` is misleading, that is used for [keepAlive type messages](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/transport/TcpTransport.java#L207), for all other messages (TransportRequest, TransportResponse etc.), the messages are sent using [`internalSend`](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/transport/OutboundHandler.java#L195). Before caling `internalSend`, the message will be [serialized](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/transport/OutboundHandler.java#L189) which should have different implementations for each protocol, hence the current code in this PR.

Reply:
@reta Let me know what you think of the above.

Reply:
I have created the draft pull request to demonstrate the idea, it makes the `OutboundHandler` context dependant (on the protocol): https://github.com/opensearch-project/OpenSearch/pull/13452

Reply:
Thank you very much @reta for the draft PR! This makes it very clear! I have incorporated all those changes in this PR :)